<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

Привет! Меня зовут Сысоева Вера.
    
Далее в файле мои комментарии ты сможешь найти в ячейках, аналогичных данной ( если рамки комментария зелёные и отмечены **#ок** - всё сделано правильно; жёлтые и отмечены **#можно лучше** - есть замечания, но не критично; красные с пометкой **#ошибка** - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения проекта.
    
Я буду поддерживать версионность комментариев, и при следующих итерациях комментарии будут выделены, как "Комментраии ревьюера - 2 (3 и.т.д.)"
    
Если у тебя есть замечания, возражения и вопросы, ты можешь писать свои комментарии, по возможности выделяй их так же цветом (например, синим), чтоб их можно было легко отделить от текста проекта. 
    
Кстати, предлагаю сразу общаться на "ты".
    
Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.
    
</div>

# Яндекс.Музыка
Оценка музыкальных предпочтений пользователей сервиса "Яндекс.Музыка" в городах Москва и Санкт-Петербург.

## Цель исследования.
Проверить три гипотезы:
1. Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.
2. В понедельник утром в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города. 
3. Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

## Обзор данных.

### Получение данных.

Загрузка данных из файла: : `yandex_music_project.csv`

In [1]:
# импорт библиотеки pandas
import pandas as pd
# чтение файла с данными и сохранение в df
df = pd.read_csv('/datasets/yandex_music_project.csv')

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Немного удобне подгружать библиотеки в отдельной ячейке проекта. Так при необходимости добавить библиотеку тебе нужно будет дополнить и перезапустить первую ячеку. Эта ячейка только перегрузит библиотеки и не перегрузит данные, т.е. твой прогресс будет сохранен
    
В целом старайся, чтобы ячейки содержали некоторую логическую часть и не страшно, если в них будет по одной строке кода
    
</div>

### Первичный осмотр данных.

Проверка загрузки и общего представления данных:

In [2]:
# получение 10 строк таблицы df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                    Преданная         IMPERVTOR  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE     И вновь продолжается бой               NaN  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

             City        time        Day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday  
5  Saint-Petersburg  13:09:41     Friday  
6            Moscow  13:00:07  Wednesday  
7            Moscow  20:47:49  Wednesday  
8            Moscow  09:17:40     Friday  
9  Saint-Petersburg  21:20:49  Wednesday

<div style="border:solid red 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ошибка**

Для просмотра первых строк надо использовать метод `head`
</div>

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера - 2**

**#ок**
    
👍

    
</div>

Просмотр общей информации по данным.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
  userID    65079 non-null object
Track       63848 non-null object
artist      57876 non-null object
genre       63881 non-null object
  City      65079 non-null object
time        65079 non-null object
Day         65079 non-null object
dtypes: object(7)
memory usage: 3.5+ MB


Первичная информация по полученным данным:
1. Загрузка данных производится корректно.
2. Данные представлены как события прослушивания трека, таких событий 61253.
3. По каждому событию имеются следующие данные (согласно представленной документации):
    * `userID` — идентификатор пользователя;
    * `Track` — название трека;  
    * `artist` — имя исполнителя;
    * `genre` — название жанра;
    * `City` — город пользователя;
    * `time` — время начала прослушивания;
    * `Day` — день недели.
4. В названиях колонок видны нарушения стиля:
    * строчные буквы сочетаются с прописными.
    * встречаются пробелы.
    * не используется "змеиный_регистр" при использовании нескольких слов.
5. Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.


<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Все проблемы в названиях правильно обозначены
</div>

### Выводы по результату обзора данных

В каждой строке таблицы — данные о прослушанном треке. Часть колонок описывает саму композицию: название, исполнителя и жанр. Остальные данные рассказывают о пользователе: из какого он города, когда он слушал музыку. 

Предварительно можно утверждать, что, данных достаточно для проверки гипотез. Но встречаются пропуски в данных, а в названиях колонок — расхождения с хорошим стилем.

## Предобработка данных

### Исправление стиля заголовков.

Приведение стиля заголовков столбцов к единому стилю "змеиный_регистр".

In [4]:
# перечень названий столбцов таблицы df
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Необходимо исправить следующие названия столбцов:
* `'  userID'` → `'user_id'`;
* `'Track'` → `'track'`;
* `'  City  '` → `'city'`;
* `'Day'` → `'day'`.

In [5]:
# переименование столбцов
df = df.rename( columns={
'  userID': 'user_id',
'Track': 'track',
'  City  ': 'city',
'Day': 'day'
 }
)

# проверка результатов - перечень названий столбцов
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

Переменование прошло успешно.

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Хорошо, проблемы с названиями решены.

</div>

### Устранение пропусков значений.

На первом этапе устраним явные пропуски значений.

In [6]:
# подсчёт пропусков
df.isna().sum()

user_id       0
track      1231
artist     7203
genre      1198
city          0
time          0
day           0
dtype: int64

Имеются пропуски в данных `track`, `artist` и `genre`.
Произведём замену пропусков явным обозначением `unknown`, при этом нужно учесть, что согласно целей исследования наши результаты зависимы от данных жанра и независимы от названий исполнителя и трека, поэтому:
    1. Замена значение в столбцах `track`, `artist` не влияют на результаты исследования;
    2. Нужно будет оценить влияние замены значений в столбце `genre` на результаты исследования.

In [7]:
# перебор названий столбцов в цикле и замена пропущенных значений на 'unknown'
columns_to_replace = ['track','artist','genre']
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')
    
# подсчёт пропусков после замены пропущёных значений
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

Устранение пропусков произведено корректно.

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Отлично, все пропуски заполнены правильно, верно использован цикл

</div>

### Устранение дубликатов

Вначале посчитаем количество явных дубликатов в данных.

In [8]:
# подсчёт явных дубликатов
df.duplicated().sum()

3826

Количесвто дубликатов равно 3826, довольно много для 65079 событий (больше 5%), нужно будет передать информацию в отдел сбора данных.
Произведём удаление явных дубликатов и проверим их отсутствие.

In [9]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df = df.drop_duplicates().reset_index(drop=True)

# проверка на отсутствие дубликатов
df.duplicated().sum()

0

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Ок, все верно. Классно разобрались с дубликатами, но стоит подписывать выводимые значения не комментариями в коде, а непосредственно при выводе. Например,
    
"Количество явных дубликатов: 3826"

</div>

Явные дубликаты удалены.
Для качественного достижения заявленных целей исследования необходимо проверить данные в колонке 'genre' на наличие неявных дубликатов, для этого нужно изучить список уникальных названий жанров (для удобства отсортированных в алфавитном порядке).

In [10]:
# Просмотр уникальных названий жанров, отсортированных в алфивитном порядке

df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Отлично, сортировка правильно сделана. Верно, что ты сделал все в одну строку, без дополнительных переменных
</div>

Визуальный осмотр таблицы позволил выявить следующие неявные дубликаты:
* *hip*, *hop*, *hip-hop* - `hiphop`,
* *ïîï* - `indie`,
* *электроника* - `electronic`,
* *türkçe* - `türk`
* *afrikaans* - `africa`

Напишем функцию для исправления неявных дубликатов `replace_wrong_genres()` с двумя параметрами: 
* `wrong_genres` — список дубликатов,
* `correct_genre` — строка с правильным значением.

In [11]:
# Функция для замены неявных дубликатов
# Функция должна исправить колонку `genre` в таблице `df`: заменить каждое значение из списка `wrong_genres` на значение из `correct_genre`

def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

Устраним выявленные неявные дублирования вызовом функции `replace_wrong_genres()`, передав ей аргументы для каждого жанра.
И выведем изменённый список жанров для контроля.

In [12]:
# список аргументов для hiphop, indie, electronic, türk, africa, передаваемых циклом в функцию replace_wrong_genres()
arguments_for_replace = [
    [['hip', 'hop', 'hip-hop'], 'hiphop'],
    [['ïîï'], 'indie'],
    [['электроника'], 'electronic'],
    [['türkçe'], 'türk'],
    [['afrikaans'], 'africa']]
    
# Циклический вызов функции для замены неявных дубликатов
for argument in arguments_for_replace:
    wrong_genre = argument[0]
    correct_genre = argument[1]
    replace_wrong_genres(wrong_genre, correct_genre)
    
# Проверка на неявные дубликаты
df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'alternative',
       'alternativepunk', 'ambient', 'americana', 'animated', 'anime',
       'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook',
       'author', 'avantgarde', 'axé', 'baile', 'balkan', 'beats',
       'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa',
       'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori',
       'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic',
       'chamber', 'chanson', 'children', 'chill', 'chinese', 'choral',
       'christian', 'christmas', 'classical', 'classicmetal', 'club',
       'colombian', 'comedy', 'conjazz', 'contemporary', 'country',
       'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep',
       'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb',
       'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep',
       'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno',
       'epicmetal', '

Визуальный осмотр таблицы подтвердил устранение неявных дубликовто в названиях жанров.

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Хорошо, ошибки в написании жанров поправили. Здорово, что ты посмотрел на проблемы шире, чем в задании. 

</div>

### Выводы по результату предобработки данных.

В ходе предобработки данных были устранены следующие замечания и нарушения:
- заголовки приведены к единому стилю "змеиный_регистр",
- заполнены отсутствующие данные на `'unknown'`,
- устранены дубликаты: явные и неявные.

Обратить особое внимание:
- передать в отдел сбора данных информацию о большом количестве дубликатов в БД, более 5 %,
- проверить влияние изначальных пропусков в колонке `'genre'` на результаты исследования.

Данные готовы для проверки гипотез обозначеных в целях исследования.

## Проверка гипотез

### Сравнение поведения пользователей двух столиц

Первая гипотеза утверждает, что пользователи по-разному слушают музыку в Москве и Санкт-Петербурге. Проверим это предположение по данным о трёх днях недели — понедельнике, среде и пятнице.

Для этого создадим функцию `number_tracks()`, которая посчитает прослушивания для заданного дня и города. Ей понадобятся два параметра:
* день недели - `day`,
* название города - `city`.

In [13]:
# <создание функции number_tracks()>
# Параметры: day - день недели, city - город.
# В переменной track_list сохраняются те строки таблицы df, для которых 
# значение в столбце 'day' равно параметру day и одновременно значение
# в столбце 'city' равно параметру city.
# В переменной track_list_count сохраняется число значений столбца 'user_id',
# рассчитанное методом count() для таблицы track_list.
# Функция возвращает число - значение track_list_count.

def number_tracks(day, city):
    track_list = df[(df['day'] == day) & (df['city'] == city)]
    track_list_count = track_list['user_id'].count()
    return track_list_count  

Создадим словарь списков `number_tracks_result`, который через два цикл заполним результатами вычислений функции `number_tracks()`.
C помощью конструктора `pd.DataFrame` создадим таблицу, параметром которого будет словарь списков `number_tracks_result`, где
* названия колонок — ключи словаря списков `number_tracks_result`;
* данные — списки к соответствующими ключами (дни недели) словаря списков `number_tracks_result`.

In [14]:
# словарь списков вычилений количества прослушиваний для дней недели и города
number_tracks_result = {'city':['Moscow','Saint-Petersburg'], 'Monday':[], 'Wednesday':[], 'Friday':[]}

# Заполнение значениями словаря списков, в первом цикле перебираются дни недели, во втором цикле перебираются города,
# значения вычисляются вызовом функции number_tracks() и заносятся в словарь списков
for day in ['Monday','Wednesday','Friday']:
    for city in number_tracks_result['city']:
        number_tracks_result[day].append(number_tracks(day, city))

#Создадим таблицу DataFrame, из словаря списков number_tracks_result и выведем на экран
df_number_tracks_result = pd.DataFrame(number_tracks_result)
df_number_tracks_result.set_index('city', inplace = True)
display(df_number_tracks_result)

Monday  Wednesday  Friday
city                                       
Moscow             15740      11056   15945
Saint-Petersburg    5614       7003    5895

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Отлично, датасет правильно собран.

Немного более точно было бы без цикла написать

    pd.DataFrame(data=[['Moscow', number_tracks(day='Monday', city='Moscow'), number_tracks(day='Wednesday', city='Moscow'), number_tracks(day='Friday', city='Moscow')], ...
                                        
    
</div>

Проверка гипотезы показывает, что имеется отличие в прослушивании музыки по дням недели. Среда для Москвы менее активный день, для Санкт-Петербурга это самый активный день.

### Вывод о поведении пользователей двух столиц в зависимости от дня недели.

Данные показывают разницу поведения пользователей:

- В Москве пик прослушиваний приходится на понедельник и пятницу, а в среду заметен спад.
- В Петербурге, наоборот, больше слушают музыку по средам. Активность в понедельник и пятницу здесь почти в равной мере уступает среде.

Значит, данные говорят в пользу первой гипотезы.

### Музыка в начале и в конце недели

Согласно второй гипотезе, утром в понедельник в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города.

Проверку данной гипотезы начнём с создания двух таблиц, в которые отдельно сохраним прослушивания треков для Москвы и Санкт-Петербурга:
* по Москве — в `moscow_general`;
* по Санкт-Петербургу — в `spb_general`.

In [15]:
# получение таблицы moscow_general из тех строк таблицы df, 
# для которых значение в столбце 'city' равно 'Moscow'
moscow_general = df[df['city'] == 'Moscow'].reset_index(drop=True)

# получение таблицы spb_general из тех строк таблицы df,
# для которых значение в столбце 'city' равно 'Saint-Petersburg'
spb_general = df[df['city'] == 'Saint-Petersburg'].reset_index(drop=True)

Создадим функцию `genre_weekday()` с четырьмя параметрами:
* таблица (датафрейм) с данными,
* день недели,
* начальная временная метка в формате 'hh:mm', 
* последняя временная метка в формате 'hh:mm'.

Функция должна вернуть информацию о топ-10 жанров тех треков, которые прослушивали в указанный день, в промежутке между двумя отметками времени.

In [16]:
# Объявление функции genre_weekday() с параметрами table, day, time1, time2,
# которая возвращает информацию о самых популярных жанрах в указанный день в
# заданное время:
# 1) в переменную genre_df сохраняются те строки переданного датафрейма table, для
#    которых одновременно:
#    - значение в столбце day равно значению аргумента day
#    - значение в столбце time больше значения аргумента time1
#    - значение в столбце time меньше значения аргумента time2
#    Используйем последовательную фильтрацию с помощью логической индексации.
# 2) сгруппируем датафрейм genre_df по столбцу genre, возьмём столбец 'user_id'
#    и посчитаем методом count() количество записей для каждого из
#    присутствующих жанров, получившийся Series запишем в переменную
#    genre_df_count
# 3) отсортируем genre_df_count по убыванию и сохраним
#    в переменную genre_df_sorted
# 4) вернём группировку ключей и значений Series из 10 первых значений genre_df_sorted,
#    это будут топ-10 популярных жанров (в указанный день, в заданное время).

def genre_weekday(table, day, time1, time2):
    genre_df = table[table['day'] == day]
    genre_df = genre_df[genre_df['time'] > time1]
    genre_df = genre_df[genre_df['time'] < time2]
    genre_df_count = genre_df.groupby('genre')['user_id'].count()
    genre_df_sorted = genre_df_count.sort_values(ascending=False)
    return zip(genre_df_sorted[:10].keys().tolist(), genre_df_sorted[:10].to_list())

Cравните результаты функции `genre_weekday()` для Москвы и Санкт-Петербурга в понедельник утром (с 7:00 до 11:00) и в пятницу вечером (с 17:00 до 23:00):

In [17]:
# создадим словарь, ключами которого будут: названия города, день недели и время, 
# а значениями список десяти самых популярных жанров вычесленных функцией genre_weekday()
dict_genre_weekday = {}

# ключи и заполнения словаря dict_genre_weekday заполняются циклическим вызовом функции genre_weekday(),
# первый цикл перебирает список с днём недели и временем проведения исследования,
# второй цикл перебирает таблицы с данными по городам.
for day_time in [['Monday', '07:00', '11:00'],['Friday', '17:00', '23:00']]:
    for city_df in [moscow_general, spb_general]:
        dict_genre_weekday[f"{city_df['city'][0]} {day_time[0]} {day_time[1]}-{day_time[2]}"] = genre_weekday(city_df, day_time[0], day_time[1], day_time[2])

# вывод DataFrame таблицы, созданной из словаря dict_genre_weekday       
pd.DataFrame(dict_genre_weekday)

Moscow Monday 07:00-11:00 Saint-Petersburg Monday 07:00-11:00  \
0                (pop, 781)                          (pop, 218)   
1              (dance, 549)                        (dance, 182)   
2         (electronic, 480)                         (rock, 162)   
3               (rock, 474)                   (electronic, 147)   
4             (hiphop, 286)                        (hiphop, 80)   
5             (ruspop, 186)                        (ruspop, 64)   
6              (world, 181)                   (alternative, 58)   
7             (rusrap, 175)                        (rusrap, 55)   
8        (alternative, 164)                          (jazz, 44)   
9            (unknown, 161)                     (classical, 40)   

  Moscow Friday 17:00-23:00 Saint-Petersburg Friday 17:00-23:00  
0                (pop, 713)                          (pop, 256)  
1               (rock, 517)                         (rock, 216)  
2              (dance, 495)                   (electronic, 216)  
3         (electronic, 482)                        (dance, 210)  
4             (hiphop, 273)                        (hiphop, 97)  
5              (world, 208)                   (alternative, 63)  
6             (ruspop, 170)                          (jazz, 61)  
7          (classical, 163)                     (classical, 60)  
8        (alternative, 163)                        (rusrap, 59)  
9             (rusrap, 142)                         (world, 54)

Из таблицы видно, что значительных расхождений в жанровых предпочтениях не наблюдается. Абсолютный лидер по жанрам `Рор`.
Влияние отсутствующих данных `unknown`, также не должно принципиально изменить расстановку жанров, наблюдается только на 10 месте для Москвы утро понедельника (может повлиять только на нижнюю часть таблицы).

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Ок, функция написана верно и результат требуемый получен

Все же немного более читаемым был бы вариант с 4-мя отдельными таблицами
</div>

### Выводы о влиянии начала и конца недели на жанровые предпочтения.

Если сравнить топ-10 жанров в понедельник утром, можно сделать такие выводы:

1. В Москве и Петербурге слушают похожую музыку. Единственное отличие — в московский рейтинг вошёл жанр “world”, а в петербургский — джаз и классика.

2. В Москве пропущенных значений оказалось достаточно, что значение `'unknown'` заняло десятое место среди самых популярных жанров. Но значительного влияния на лидирующие жанры не оказывает.

Вечер пятницы не меняет эту картину. Некоторые жанры поднимаются немного выше, другие спускаются, но в целом топ-10 остаётся тем же самым.

Таким образом, вторая гипотеза подтвердилась лишь частично:
* Пользователи слушают похожую музыку в начале недели и в конце.
* Разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую популярную музыку, в Петербурге — джаз.

### Жанровые предпочтения в Москве и Петербурге



Третья гипотеза: Петербург — столица рэпа, музыку этого жанра там слушают чаще, чем в Москве.  А Москва — город контрастов, в котором, тем не менее, преобладает поп-музыка.

Сгруппируем таблицу `moscow_general` по жанру и посчитаем прослушивания треков каждого жанра методом `count()`. Затем отсортируем результат в порядке убывания и сохраним его в таблице `moscow_genres` и выведем первые 10 строк данной таблицы.

In [18]:
# одной строкой: группировка таблицы moscow_general по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в moscow_genres

moscow_genres = moscow_general.groupby('genre')['genre'].count().sort_values(ascending=False)
moscow_genres[:10]

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

Повторим то же для Санкт-Петербурга.

Сгруппируем таблицу `spb_general` по жанру и посчитаем прослушивания треков каждого жанра методом `count()`. Затем отсортируем результат в порядке убывания и сохраним его в таблице `spb_genres` и выведем первые 10 строк данной таблицы.

In [19]:
# одной строкой: группировка таблицы spb_general по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в spb_genres

spb_genres = spb_general.groupby('genre')['genre'].count().sort_values(ascending=False)
spb_genres[:10]

genre
pop            2431
dance          1932
rock           1879
electronic     1737
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

Для удобста сравнения результатов по Москве и Санкт-Петербургу, отобразим результаты по жанровым предпочтениям в одной таблице.
При помощи метода `zip` сгруппируем ключи и значения таблиц `moscow_genres` и `spb_genres`, результат группировки сохраним в переменной `moskov_spb_genres`.
На основе переменной `moskov_spb_genres` сформируем таблицу предпочтения по первым 10 жанрам для Москвы и Санкт-Петербурга и выведем её на экран.

In [20]:
# методом zip() производится три группировки, результат соханяется в переменной moskov_spb_genres:
# 1) группируются жанры и значения прослушиваний для них из таблицы moscow_genres (10 самых популярных),
# 2) группируются жанры и значения прослушиваний для них из таблицы spb_genres (10 самых популярных),
# 3) группируются результаты первых двух группирововк, что в дальнейшем будет использована как строки в DataFrame

moskov_spb_genres = zip(zip(moscow_genres[:10][:10].keys().tolist(),
                            moscow_genres[:10][:10].to_list()),
                        zip(spb_genres[:10][:10].keys().tolist(),
                            spb_genres[:10][:10].to_list()))

# переменная moskov_spb_genres выступает параметром для конструктора DataFrame, результат выводится на экран

pd.DataFrame(moskov_spb_genres, columns=('Moscow','Saint-Petersburg'))

Moscow    Saint-Petersburg
0          (pop, 5892)         (pop, 2431)
1        (dance, 4435)       (dance, 1932)
2         (rock, 3965)        (rock, 1879)
3   (electronic, 3786)  (electronic, 1737)
4       (hiphop, 2096)       (hiphop, 960)
5    (classical, 1616)  (alternative, 649)
6        (world, 1432)    (classical, 646)
7  (alternative, 1379)       (rusrap, 564)
8       (ruspop, 1372)       (ruspop, 538)
9       (rusrap, 1161)        (world, 515)

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Отлично, в этой части тоже все верно сделано
</div>

### Выводы о жанровых предпочтениях.

Гипотеза о жанровых предпочтениях двух городов частично подтвердилась:
* Поп-музыка — самый популярный жанр в Москве, как и предполагала гипотеза. Более того, в топ-10 жанров встречается близкий жанр — русская популярная музыка.
* Вопреки ожиданиям, рэп в Санкт-Петербурге занимает только 8 местов рейтинге, а возглавляет рейтинг, так-же как и в Мосвке популярная музыка. 


## Итоги исследования

**В ходе исследования была проведена проверка трёх гипотез и установлено:**

1. День недели по-разному влияет на активность пользователей в Москве и Петербурге. Установлено:
    * В Москве пик прослушиваний приходится на понедельник и пятницу, а в среду заметен спад.
    * В Петербурге, наоборот, больше слушают музыку по средам. Активность в понедельник и пятницу здесь почти в равной мере уступает среде.


2. Музыкальные предпочтения не сильно меняются в течение недели — будь то Москва или Петербург. Небольшие различия заметны в начале недели, по понедельникам:
   * Пользователи слушают похожую музыку в начале недели и в конце.
   * Разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую популярную музыку, в Петербурге — джаз.


3. Во вкусах пользователей Москвы и Петербурга больше общего чем различий. Вопреки ожиданиям, предпочтения жанров в Петербурге напоминают московские:
    * Поп-музыка — самый популярный жанр в Москве, как и предполагала гипотеза. Более того, в топ-10 жанров встречается близкий жанр — русская популярная музыка.
    * Вопреки ожиданиям, рэп в Санкт-Петербурге занимает только 8 местов рейтинге, а возглавляет рейтинг, так-же как и в Мосвке популярная музыка. 


**Дополнительная информация:**
* в ходе предобработки данных установлено наличие прокусков в данных, это не повлияло значительно на результаты исследования, но рекомендуется принять меры для более корректного сбора данных,
* в ходе предобработки данных установлено наличие большого количества дубликатов (более 5%), рекомендуется принять меры для более корректного сбора данных.

<div style="border:solid green 2px; padding: 20px"> 

**Общий комментарий по проекту**

Ты отлично поработал над проектом, молодец. Ты сделал хороший самостоятельный анализ.
    
Ты молодец, отлично справляешься с программой.  Дальше все будет интереснее и сложнее.
    
Поправь, пожалуйста, небольшую неточность, чтобы я могла принять проект.

Удачи.
    
</div>